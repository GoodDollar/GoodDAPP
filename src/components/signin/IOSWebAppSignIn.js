// @flow
//eslint-disable-next-line

import bip39 from 'bip39-light'
import React, { Fragment, useState } from 'react'
import { AsyncStorage, Image } from 'react-native'
import { IS_LOGGED_IN } from '../../lib/constants/localStorage'
import logger, { ExceptionCategory } from '../../lib/logger/pino-logger'
import { withStyles } from '../../lib/styles'
import { useErrorDialog } from '../../lib/undux/utils/dialog'
import retryImport from '../../lib/utils/retryImport'
import Text from '../common/view/Text'
import Section from '../common/layout/Section'
import CustomButton from '../common/buttons/CustomButton'
import InputText from '../common/form/InputText'
import NavBar from '../appNavigation/NavBar'
import IOSWebAppSignInSVG from '../../assets/IOSWebAppSignIn.svg'
import { getDesignRelativeHeight } from '../../lib/utils/sizes'

Image.prefetch(IOSWebAppSignInSVG)

const TITLE = 'EASY ACCESS'
const log = logger.child({ from: 'IOS EASY ACCESS' })

const IOSWebAppSignIn = ({ screenProps, navigation, styles }) => {
  //lazy load heavy wallet stuff for fast initial app load (part of initial routes)
  const [code, setCode] = useState()
  const [isRecovering, setRecovering] = useState(false)
  const [isValid, setValid] = useState(false)
  const [showErrorDialog] = useErrorDialog()

  const handleChange = (newCode: string) => {
    log.info({ newCode })

    setCode(newCode)

    if (newCode) {
      setValid(true)
    } else {
      setValid(false)
    }
  }

  const recover = async () => {
    setRecovering(true)

    const errorText = 'You are using wrong sign in code'
    let userNameAndPWD = Buffer.from(code, 'base64').toString('ascii')
    let userNameAndPWDArray = userNameAndPWD.split('+')

    log.debug('IOSSignInByMagicCode', { code, userNameAndPWDArray })

    if (userNameAndPWDArray.length === 2) {
      const userName = userNameAndPWDArray[0]
      const userPwd = userNameAndPWDArray[1]
      const UserStorage = await retryImport(() => import('../../lib/gundb/UserStorageClass').then(_ => _.UserStorage))

      const mnemonic = await UserStorage.getMnemonic(userName, userPwd)

      if (mnemonic && bip39.validateMnemonic(mnemonic)) {
        const mnemonicsHelpers = retryImport(() => import('../../lib/wallet/SoftwareWalletProvider'))
        const { saveMnemonics } = await mnemonicsHelpers

        await saveMnemonics(mnemonic)
        await AsyncStorage.setItem(IS_LOGGED_IN, true)

        window.location = '/'
      } else {
        log.error(errorText, '', null, {
          code,
          category: ExceptionCategory.Human,
          dialogShown: true,
        })
        showErrorDialog(errorText)
      }
    } else {
      log.error(errorText, '', null, {
        code,
        category: ExceptionCategory.Human,
        dialogShown: true,
      })
      showErrorDialog(errorText)
    }

    setRecovering(false)
  }

  const handleEnter = async (event: { nativeEvent: { key: string } }) => {
    if (event.nativeEvent.key === 'Enter' && isValid) {
      await recover()
    }
  }

  return (
    <Fragment>
      <NavBar title={TITLE} />
      <Section grow style={styles.wrapper}>
        <Section.Stack grow style={styles.instructions} justifyContent="center">
          <Text fontWeight="medium" fontSize={22} fontFamily="Roboto" lineHeight={28} style={styles.headerText}>
            {'Easy access with iPhone!'}
          </Text>
          <Text color="gray80Percent" fontSize={14}>
            {'Copy-paste the easy access code\nsent to you by text message:'}
          </Text>
        </Section.Stack>
        <Section.Stack justifyContent="space-between">
          <Section.Row justifyContent="center">
            <InputText value={code} onChangeText={handleChange} onKeyPress={handleEnter} showCleanAdornment autoFocus />
          </Section.Row>
        </Section.Stack>
        <Image source={IOSWebAppSignInSVG} resizeMode={'contain'} style={styles.image} />
        <Section style={styles.bottomContainer}>
          <CustomButton style={styles.buttonLayout} onPress={recover} disabled={!isValid || isRecovering}>
            {'SIGN IN'}
          </CustomButton>
        </Section>
      </Section>
    </Fragment>
  )
}

IOSWebAppSignIn.navigationOptions = {
  title: TITLE,
}

const mnemonicsStyles = ({ theme }) => ({
  headerText: {
    marginBottom: getDesignRelativeHeight(5),
  },
  wrapper: {
    borderRadius: 0,
    paddingVertical: 0,
  },
  instructions: {
    marginTop: getDesignRelativeHeight(10),
  },
  buttonLayout: {
    marginTop: getDesignRelativeHeight(20),
    marginBottom: getDesignRelativeHeight(20),
  },
  bottomContainer: {
    maxHeight: getDesignRelativeHeight(100),
    minHeight: getDesignRelativeHeight(100),
    justifyContent: 'flex-end',
  },
  image: {
    flexGrow: 1,
    flexShrink: 0,
    marginBottom: getDesignRelativeHeight(20),
    height: getDesignRelativeHeight(200),
  },
})

export default withStyles(mnemonicsStyles)(IOSWebAppSignIn)
