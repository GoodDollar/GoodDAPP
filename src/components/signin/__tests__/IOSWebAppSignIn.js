import React from 'react'
import renderer from 'react-test-renderer'
import IOSWebAppSignIn from '../IOSWebAppSignIn'
import { withThemeProvider } from '../../../__tests__/__util__'

const IOSWebAppSignInTheme = withThemeProvider(IOSWebAppSignIn)

describe('SignIn - IOSWebAppSignIn', () => {
  it('renders without errors', () => {
    const tree = renderer.create(<IOSWebAppSignInTheme />)
    expect(tree.toJSON()).toBeTruthy()
  })

  it('matches snapshot', () => {
    const component = renderer.create(<IOSWebAppSignInTheme />)
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })
})
