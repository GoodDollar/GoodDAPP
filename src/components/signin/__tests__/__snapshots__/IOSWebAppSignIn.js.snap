// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn - IOSWebAppSignIn matches snapshot 1`] = `
<div
  className="css-view-1dbjc4n r-flex-13awgt0 r-pointerEvents-12vffkv"
>
  <div
    className="css-view-1dbjc4n r-alignItems-1awozwy r-flexDirection-18u37iz r-flexShrink-1q142lx"
    style={
      Object {
        "WebkitBoxFlex": 0,
        "WebkitFlexGrow": 0,
        "backgroundColor": "rgba(0,175,255,1.00)",
        "boxShadow": "none, 0px 3px 4px rgba(0,0,0,0.24)",
        "flexGrow": 0,
        "height": "56px",
        "msFlexPositive": 0,
        "paddingLeft": "4px",
        "paddingRight": "4px",
      }
    }
  >
    <div
      className="css-view-1dbjc4n r-cursor-1loqt21 r-flex-13awgt0 r-paddingHorizontal-1e081e0 r-touchAction-1otgn73"
      data-focusable={true}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      tabIndex="0"
    >
      <h1
        className="css-reset-4rbku5 css-text-901oao css-textOneLine-bfa6kz r-color-jwli3a r-fontSize-ubezar r-fontWeight-majxgm r-textAlign-q4m81j r-textTransform-tsynxw"
        dir="auto"
        role="heading"
        style={
          Object {
            "fontFamily": "Roboto, \\"Helvetica Neue\\", Helvetica, Arial, sans-serif",
          }
        }
      >
        EASY ACCESS
      </h1>
    </div>
  </div>
  <div
    className="css-view-1dbjc4n r-backgroundColor-14lw9ot r-borderRadius-17gur6a r-flexGrow-16y2uox r-paddingHorizontal-1e081e0 r-paddingVertical-rjfia"
  >
    <div
      className="css-view-1dbjc4n r-alignItems-1oszu61 r-flexDirection-eqz5dr r-flexGrow-16y2uox r-justifyContent-1777fci r-marginTop-156q2ks"
    >
      <div
        className="css-text-901oao r-color-cddv6f r-fontFamily-vnw8o6 r-fontSize-evnaw r-fontWeight-majxgm r-lineHeight-vrz42v r-marginBottom-d0pm55 r-textAlign-q4m81j r-textDecorationLine-13wfysu r-textTransform-3twk1y"
        dir="auto"
      >
        Easy access with iPhone!
      </div>
      <div
        className="css-text-901oao r-color-vaa0vc r-fontFamily-vnw8o6 r-fontSize-1b43r93 r-fontWeight-1it3c9n r-lineHeight-rjixqe r-textAlign-q4m81j r-textDecorationLine-13wfysu r-textTransform-3twk1y"
        dir="auto"
      >
        Copy-paste the easy access code
sent to you by text message:
      </div>
    </div>
    <div
      className="css-view-1dbjc4n r-alignItems-1oszu61 r-flexDirection-eqz5dr r-justifyContent-1wtj0ep"
    >
      <div
        className="css-view-1dbjc4n r-alignItems-1awozwy r-flexDirection-18u37iz r-justifyContent-1777fci"
      >
        <div
          className="css-view-1dbjc4n r-marginBottom-5oul0u r-width-13qz1uu"
        >
          <div
            className="css-view-1dbjc4n r-marginBottom-5oul0u r-width-13qz1uu"
          >
            <input
              autoCapitalize="sentences"
              autoComplete="on"
              autoCorrect="on"
              autoFocus={true}
              className="css-textinput-1cwyjr8 r-backgroundColor-14lw9ot r-borderBottomColor-1ny7kx0 r-borderBottomWidth-qklmqi r-color-cddv6f r-fontFamily-6jnb14 r-fontSize-1i10wst r-paddingHorizontal-1dye5f7 r-paddingVertical-oyd9sg r-placeholderTextColor-1xktb6b r-textAlign-q4m81j"
              data-focusable={true}
              dir="auto"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onSelect={[Function]}
              onTouchStart={[Function]}
              readOnly={false}
              spellCheck={true}
              type="text"
            />
            <div
              className="css-view-1dbjc4n r-bottom-1ebgqk7 r-cursor-1loqt21 r-paddingTop-tskmnb r-position-u8s1d r-right-3mc0re r-touchAction-1otgn73 r-transitionDuration-eafdt9 r-transitionProperty-1i6wzkk r-userSelect-lrvibr r-zIndex-184en5c"
              data-focusable={true}
              disabled={false}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              tabIndex="0"
            >
              <div
                className="css-text-901oao r-color-cddv6f r-fontSize-ubezar r-fontWeight-1it3c9n"
                dir="auto"
                style={
                  Object {
                    "fontFamily": "gooddollar",
                    "fontStyle": "normal",
                  }
                }
              >
                î Œ
              </div>
            </div>
          </div>
          <div
            className="css-text-901oao r-paddingHorizontal-1e081e0 r-paddingVertical-oyd9sg r-textAlign-q4m81j"
            dir="auto"
            style={
              Object {
                "WebkitTransform": "translateY(0px)",
                "color": "rgba(250,108,119,1.00)",
                "direction": "ltr",
                "fontFamily": "Roboto, \\"Helvetica Neue\\", Helvetica, Arial, sans-serif",
                "fontSize": "12px",
                "fontWeight": "400",
                "height": "18px",
                "opacity": 0,
                "paddingBottom": "0px",
                "paddingLeft": "0px",
                "paddingTop": "0px",
                "transform": "translateY(0px)",
              }
            }
          />
        </div>
      </div>
    </div>
    <div
      className="css-view-1dbjc4n r-flexGrow-16y2uox r-flexShrink-1q142lx r-height-4d76ec r-marginBottom-117bsoe"
    >
      <IOSWebAppSignIn.svg />
    </div>
    <div
      className="css-view-1dbjc4n r-backgroundColor-14lw9ot r-borderRadius-1f0042m r-justifyContent-17s6mgv r-maxHeight-qqzlmc r-minHeight-h3f8nf r-paddingHorizontal-1e081e0 r-paddingVertical-1yzf0co"
    >
      <div
        className="css-view-1dbjc4n r-alignItems-1oszu61 r-display-6koalj r-justifyContent-1777fci r-marginBottom-117bsoe"
        style={
          Object {
            "backgroundColor": "rgba(0,0,0,0.12)",
            "borderBottomColor": "rgba(0,175,255,1.00)",
            "borderBottomLeftRadius": "50px",
            "borderBottomRightRadius": "50px",
            "borderBottomStyle": "solid",
            "borderBottomWidth": "0px",
            "borderLeftColor": "rgba(0,175,255,1.00)",
            "borderLeftStyle": "solid",
            "borderLeftWidth": "0px",
            "borderRightColor": "rgba(0,175,255,1.00)",
            "borderRightStyle": "solid",
            "borderRightWidth": "0px",
            "borderTopColor": "rgba(0,175,255,1.00)",
            "borderTopLeftRadius": "50px",
            "borderTopRightRadius": "50px",
            "borderTopStyle": "solid",
            "borderTopWidth": "0px",
            "marginLeft": "0px",
            "marginRight": "0px",
            "marginTop": "20px",
            "minHeight": "44px",
            "minWidth": "64px",
            "paddingBottom": "0px",
            "paddingLeft": "0px",
            "paddingRight": "0px",
            "paddingTop": "0px",
          }
        }
      >
        <div
          aria-disabled={true}
          className="css-view-1dbjc4n r-flex-13awgt0 r-overflow-1udh08x r-position-bnwqim"
          disabled={true}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          role="button"
          style={
            Object {
              "WebkitAlignSelf": "stretch",
              "alignSelf": "stretch",
              "borderBottomLeftRadius": "50px",
              "borderBottomRightRadius": "50px",
              "borderTopLeftRadius": "50px",
              "borderTopRightRadius": "50px",
              "msFlexItemAlign": "stretch",
            }
          }
        >
          <div
            className="css-view-1dbjc4n r-alignItems-1awozwy r-flex-13awgt0 r-flexDirection-18u37iz r-justifyContent-1777fci"
            style={
              Object {
                "WebkitAlignSelf": "stretch",
                "alignSelf": "stretch",
                "msFlexItemAlign": "stretch",
              }
            }
          >
            <div
              className="css-view-1dbjc4n r-marginHorizontal-1ydw1k6 r-marginVertical-712mwr"
            >
              <div
                className="css-view-1dbjc4n r-alignItems-1awozwy r-display-6koalj r-flex-13awgt0 r-flexDirection-18u37iz"
              >
                <div
                  className="css-text-901oao r-alignItems-1awozwy r-color-jwli3a r-fontFamily-vnw8o6 r-fontSize-ubezar r-fontWeight-majxgm r-justifyContent-1777fci r-letterSpacing-oxtfae r-lineHeight-hbpseb r-paddingTop-njp1lv r-textAlign-q4m81j r-textDecorationLine-13wfysu r-textTransform-tsynxw"
                  dir="auto"
                >
                  SIGN IN
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
